src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -550,10 +550,102 @@ inline bool ShenandoahHeap::clear_old_evacuation_failure() {
+  if (!mode()->is_generational()) {
+    // everything is the same single generation
+    return true;
+  }
+
+  if (active_generation() == NULL) {
+    // no collection is happening, only expect this to be called
+    // when concurrent processing is active, but that could change
+    return false;
+  }
+
+  assert(is_in(obj), "only check if is in active generation for objects (" PTR_FORMAT ") in heap", p2i(obj));
+  assert((active_generation() == (ShenandoahGeneration*) old_generation()) ||
+         (active_generation() == (ShenandoahGeneration*) young_generation()) ||
+         (active_generation() == global_generation()), "Active generation must be old, young, or global");
+
+  size_t index = heap_region_containing(obj)->index();
+  switch (_affiliations[index]) {
+  case ShenandoahRegionAffiliation::FREE:
+    // Free regions are in Old, Young, Global
+    return true;
+  case ShenandoahRegionAffiliation::YOUNG_GENERATION:
+    // Young regions are in young_generation and global_generation, not in old_generation
+    return (active_generation() != (ShenandoahGeneration*) old_generation());
+  case ShenandoahRegionAffiliation::OLD_GENERATION:
+    // Old regions are in old_generation and global_generation, not in young_generation
+    return (active_generation() != (ShenandoahGeneration*) young_generation());
+  default:
+    assert(false, "Bad affiliation (%d) for region " SIZE_FORMAT, _affiliations[index], index);
+    return false;
+  }
+}
+
+inline bool ShenandoahHeap::is_in_young(const void* p) const {
+  return is_in(p) && (_affiliations[heap_region_index_containing(p)] == ShenandoahRegionAffiliation::YOUNG_GENERATION);
+}
+
+inline bool ShenandoahHeap::is_in_old(const void* p) const {
+  return is_in(p) && (_affiliations[heap_region_index_containing(p)] == ShenandoahRegionAffiliation::OLD_GENERATION);
+}
+
 inline bool ShenandoahHeap::is_old(oop obj) const {
   return is_gc_generation_young() && is_in_old(obj);
 }
 
+inline ShenandoahRegionAffiliation ShenandoahHeap::region_affiliation(const ShenandoahHeapRegion *r) {
+  return (ShenandoahRegionAffiliation) _affiliations[r->index()];
+}
+
+inline void ShenandoahHeap::assert_lock_for_affiliation(ShenandoahRegionAffiliation orig_affiliation,
+                                                        ShenandoahRegionAffiliation new_affiliation) {
+  // A lock is required when changing from FREE to NON-FREE.  Though it may be possible to elide the lock when
+  // transitioning from in-use to FREE, the current implementation uses a lock for this transition.  A lock is
+  // not required to change from YOUNG to OLD (i.e. when promoting humongous region).
+  //
+  //         new_affiliation is:     FREE   YOUNG   OLD
+  //  orig_affiliation is:  FREE      X       L      L
+  //                       YOUNG      L       X
+  //                         OLD      L       X      X
+  //  X means state transition won't happen (so don't care)
```

Comment:
Nits:
1. Instead of the blank entry in the cell for Young->Old, may be use `N` and replace `Blank` in the legend below?
2. If we don't expect some transitions (and believe those are incorrect), it might make sense to assert that in `set_affiliation()`?
3. Further below, it looks like all current (and expected) uses of `assert_lock_for_affiliation` will be from assertion checking code in non-release builds. In that case, you can avoid the `#ifdef ASSERT` verbage at call sites by declaring the method as `PRODUCT_RETURN` which should elide the empty method in release builds.